editor/plugins/animation_library_editor.cpp

```diff
@@ -0,0 +1,689 @@
+		add_library_validate->set_text(error);
+		add_library_dialog->get_ok_button()->set_disabled(true);
+	} else {
+		add_library_validate->add_theme_color_override("font_color", get_theme_color(SNAME("success_color"), SNAME("Editor")));
+		if (name == "") {
+			add_library_validate->set_text(TTR("Global library will be created."));
+		} else {
+			add_library_validate->set_text(TTR("Library name is valid."));
+		}
+		add_library_dialog->get_ok_button()->set_disabled(false);
+	}
+}
+
+void AnimationLibraryEditor::_add_library_confirm() {
+	if (adding_animation) {
+		String anim_name = add_library_name->get_text();
+		UndoRedo *undo_redo = EditorNode::get_singleton()->get_undo_redo();
+
+		Ref<AnimationLibrary> al = player->call("get_animation_library", adding_animation_to_library);
+		ERR_FAIL_COND(!al.is_valid());
+
+		Ref<Animation> anim;
+		anim.instantiate();
+
+		undo_redo->create_action(vformat(TTR("Add Animation to Library: %s"), anim_name));
+		undo_redo->add_do_method(al.ptr(), "add_animation", anim_name, anim);
+		undo_redo->add_undo_method(al.ptr(), "remove_animation", anim_name);
+		undo_redo->add_do_method(this, "_update_editor", player);
+		undo_redo->add_undo_method(this, "_update_editor", player);
+		undo_redo->commit_action();
+
+	} else {
+		String lib_name = add_library_name->get_text();
+		UndoRedo *undo_redo = EditorNode::get_singleton()->get_undo_redo();
+
+		Ref<AnimationLibrary> al;
+		al.instantiate();
+
+		undo_redo->create_action(vformat(TTR("Add Animation Library: %s"), lib_name));
+		undo_redo->add_do_method(player, "add_animation_library", lib_name, al);
+		undo_redo->add_undo_method(player, "remove_animation_library", lib_name);
+		undo_redo->add_do_method(this, "_update_editor", player);
+		undo_redo->add_undo_method(this, "_update_editor", player);
+		undo_redo->commit_action();
+	}
+}
+
+void AnimationLibraryEditor::_load_library() {
+	List<String> extensions;
+	ResourceLoader::get_recognized_extensions_for_type("AnimationLibrary", &extensions);
+
+	file_dialog->set_title("Load Animation");
+	file_dialog->clear_filters();
+	for (const String &K : extensions) {
+		file_dialog->add_filter("*." + K);
+	}
+
+	file_dialog->set_file_mode(EditorFileDialog::FILE_MODE_OPEN_FILE);
+	file_dialog->set_current_file("");
+	file_dialog->popup_centered_ratio();
+
+	file_dialog_action = FILE_DIALOG_ACTION_OPEN_LIBRARY;
+}
+
+void AnimationLibraryEditor::_file_popup_selected(int p_id) {
+	Ref<AnimationLibrary> al = player->call("get_animation_library", file_dialog_library);
+	Ref<Animation> anim;
+	if (file_dialog_animation != StringName()) {
+		anim = al->get_animation(file_dialog_animation);
+		ERR_FAIL_COND(anim.is_null());
+	}
+	switch (p_id) {
+		case FILE_MENU_SAVE_LIBRARY: {
+			if (al->get_path().is_resource_file()) {
+				EditorNode::get_singleton()->save_resource(al);
+				break;
+			}
+		}
+			[[fallthrough]];
```

Comment:
Both styles work, but this one is more commonly used.
```suggestion
		case FILE_MENU_SAVE_LIBRARY: {
			if (al->get_path().is_resource_file()) {
				EditorNode::get_singleton()->save_resource(al);
				break;
			}
			[[fallthrough]];
		}
```